File: c/eu012.c
// project euler (projecteuler.net) problem 12
// solution by Kevin Retzke (retzkek@gmail.com)
// November 2012

#include <stdio.h>
#include <stdbool.h>

#include "pe.h"

int triangle(int n)
{
	int t = 0;
	for (int i = 1; i <= n; i++)
		t += i;
	return t;
}

int nFactors(int n)
{
	int r, i;
	r = 0;
	i = 1;
	while (i*i < n) {
		if (n%i == 0)
			r += 2;
		i++;
	}
	return r;
}

int firstOverNFactors(int n)
{
	int i, t, tf;
	i = 1;
	while (true) {
		t = triangle(i);
		tf = nFactors(t);
		if (tf > n) {
			break;
		}
		i++;
	}
	return t;
}

int main()
{
	pe_test_eq(triangle(7), 28, "%d");
	pe_test_eq(firstOverNFactors(5), 28, "%d");

	printf("%d\n",firstOverNFactors(500));
}

File: go/eu012.go
// project euler (projecteuler.net) problem 12
// solution by Kevin Retzke (retzkek@gmail.com)
// November 2012
package main

import (
	"fmt"
	"os"
)

func triangle(n int) int {
	t := 0
	for i := 1; i <= n; i++ {
		t += i
	}
	return t
}

func nFactors(n int) int {
	r := 0
	i := 1
	for i*i < n {
		if n%i == 0 {
			r += 2
		}
		i++
	}
	return r
}

func firstOverNFactors(n int) int {
	var t int
	for i := 0; ; i++ {
		t = triangle(i)
		tf := nFactors(t)
		if tf > n {
			break
		}
	}
	return t
}

func main() {
	if triangle(7) != 28 {
		fmt.Printf("Test failed. triangle(7) = %d (should be 28)\n", triangle(7))
		os.Exit(1)
	}
	if firstOverNFactors(5) != 28 {
		fmt.Printf("Test failed. firstOverNFactors(5) = %d (should be 28)\n",
			firstOverNFactors(5))
		os.Exit(1)
	}
	fmt.Println(firstOverNFactors(500))
}
